HDDS-10452. Improve Recon Disk Usage to fetch and display Top N records based on size.

ArafatKhan2198 opened the pull request:
## What changes were proposed in this pull request?
- This pull request introduces enhancements to the Recon disk usage endpoint to significantly improve usability and performance when dealing with large datasets:
- **Top Entities Focus**: The endpoint has been updated to efficiently sort and display only the top entities by size. This targeted approach helps users easily identify the most significant space consumers, addressing the impracticality of visualizing thousands of records in a single view.
- **Efficient Sorting with Parallel Streams**: To manage and sort vast numbers of records effectively, we've implemented parallel stream processing. 
- Key advantages of using parallel streams include :- 
  1. **Better Utilization of Multi-core Processors**: Enables concurrent sorting operations across multiple cores, drastically cutting down processing times for large datasets.
  2. **Optimized for Large Datasets**: The parallelism overhead is more efficiently distributed over a large number of elements, making it particularly suited for our use case.
  

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-10452

## How was this patch tested?
### Manually Tested Out the API and also using Integration Testing :- 
### Results from Manual Testing :- 
- Created 4 files of `100MB`, `10MB`, `1MB` & `10KB` under `dir-1`
```
{
  "status": "OK",
  "path": "/volumetest/buckettest/dir1",
  "size": 111010000,
  "sizeWithReplica": -1,
  "subPathCount": 4,
  "subPaths": [
    {
      "key": true,
      "path": "/volumetest/buckettest/dir1/key100MB",
      "size": 100000000,
      "sizeWithReplica": -1,
      "isKey": true
    },
    {
      "key": true,
      "path": "/volumetest/buckettest/dir1/key10mb",
      "size": 10000000,
      "sizeWithReplica": -1,
      "isKey": true
    },
    {
      "key": true,
      "path": "/volumetest/buckettest/dir1/key1MB",
      "size": 1000000,
      "sizeWithReplica": -1,
      "isKey": true
    },
    {
      "key": true,
      "path": "/volumetest/buckettest/dir1/key10kb",
      "size": 10000,
      "sizeWithReplica": -1,
      "isKey": true
    }
  ],
  "sizeDirectKey": 111010000
}

```

SaketaChalamchala commented:
@devmadhuu and @dombizita could you please take a look?

smitajoshi12 commented:
@ArafatKhan2198 
Arfafat Can you set Limit on API

devmadhuu commented on hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/api/TestNSSummaryDiskUsageOrdering.java:
```diff
@@ -0,0 +1,421 @@
+  private static final String TEST_USER = "TestUser";
+  private OMMetadataManager omMetadataManager;
+  @BeforeEach
+  public void setUp() throws Exception {
```
Is this all OM data to be inserted before each test ? In some tests we have limited paths.

devmadhuu commented on hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/api/TestNSSummaryDiskUsageOrdering.java:
```diff
@@ -0,0 +1,421 @@
+    return objectId;
+  }
+
+  private static ReconStorageContainerManagerFacade getMockReconSCM()
```
You can remove throws here. I think it doesn't throw the exception.

devmadhuu commented on hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/api/TestNSSummaryDiskUsageOrdering.java:
```diff
@@ -0,0 +1,421 @@
+    return reconSCM;
+  }
+
+  private static BucketLayout getBucketLayout() {
```
Do we have tests for LEGACY with FS Enabled ?

ArafatKhan2198 replied:
We do not have to worry about the bucket layout, the sorting happens after the response is created, hence its independent of any layouts.

devmadhuu replied:
Since we have touched size calculation in all entity handler classes, this should be tested for LEGACY and OBS layout as well.

devmadhuu commented on hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/api/TestNSSummaryDiskUsageOrdering.java:
```diff
@@ -0,0 +1,421 @@
+  private void verifyOrdering(String path)
+      throws IOException {
+    Response response =
+        nsSummaryEndpoint.getDiskUsage(path, true, false);
```
Is this API will be returning all data ? or this PR is only to just sort data in du response object ?
Because this PR title says that to display Top N records.

ArafatKhan2198 replied:
I have added a limit parameter to return only the top N disk usage records.
Please check the value is placed inside the constant :- `DISK_USAGE_TOP_RECORDS_LIMIT`

devmadhuu reviewed (commented):
Thanks @ArafatKhan2198 for working on this patch. Few comments.

ArafatKhan2198 commented:
@devmadhuu @adoroszlai @smitajoshi12

Could you please review the latest changes? Here's a quick summary:

- Switched to Parallel Sorting: To improve performance, we're now using parallel sorting. More details are in the description.
- Added a Toggle for Sorting: There's a new boolean flag to turn sorting on or off.
- Set a Limit of 30 Records: We've added a constant to limit the response to the top 30 records in Disk Usage.

dombizita reviewed (commented):
Thanks for working on this @ArafatKhan2198, overall it looks good to me, I'd like to make the javadoc and comments more accurate, please see my comments inline.

ArafatKhan2198 commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java:
```diff
@@ -101,17 +101,17 @@ public Response getBasicInfo(
    * @param path request path
    * @param listFile show subpath/disk usage for each key
    * @param withReplica count actual DU with replication
+   * @param sort whether to sort the subpaths by their sizes in descending order
```
Thanks for the input, I have made the changes!

ArafatKhan2198 commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/handlers/BucketEntityHandler.java:
```diff
@@ -142,7 +145,15 @@ public DUResponse getDuResponse(
     duResponse.setSize(bucketDataSize);
+
+    if (sortSubpaths) {
+      // Parallel sort directory/files DU data in descending order of size
```
Thanks for the input, I have made the changes!

ArafatKhan2198 commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/handlers/DirectoryEntityHandler.java:
```diff
@@ -154,6 +157,13 @@ public DUResponse getDuResponse(
     duResponse.setSize(dirDataSize);
+
+    if (sortSubPaths) {
+      // Parallel sort subdirDUData in descending order of size
```
Thanks for the input, I have made the changes!

ArafatKhan2198 commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/handlers/RootEntityHandler.java:
```diff
@@ -137,6 +140,13 @@ public DUResponse getDuResponse(
     duResponse.setSize(totalDataSize);
+
+    if (sortSubPaths) {
+      // Parallel sort volumeDuData in descending order of size
```
Thanks for the input, I have made the changes!

ArafatKhan2198 commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/handlers/VolumeEntityHandler.java:
```diff
@@ -131,6 +135,13 @@ public DUResponse getDuResponse(
     duResponse.setSize(volDataSize);
+
+    if (sortSubPaths) {
+      // Parallel sort bucketDuData in descending order of size
```
Thanks for the input, I have made the changes!

devmadhuu commented:
> @devmadhuu @adoroszlai @smitajoshi12
> 
> Could you please review the latest changes? Here's a quick summary:
> 
>     * Switched to Parallel Sorting: To improve performance, we're now using parallel sorting. More details are in the description.
> 
>     * Added a Toggle for Sorting: There's a new boolean flag to turn sorting on or off.
> 
>     * Set a Limit of 30 Records: We've added a constant to limit the response to the top 30 records in Disk Usage.

Thanks @ArafatKhan2198 for handling some points. However I am not sure if parallelStreaming always improves performance, in fact rather sometimes, it increases more overhead and may do bad than good. I would like you to have a look [here](https://blogs.oracle.com/javamagazine/post/java-parallel-streams-performance-benchmark).

devmadhuu reviewed (commented):
Some comments are still open. Pls handle them.

devmadhuu commented on hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/api/TestNSSummaryDiskUsageOrdering.java:
```diff
@@ -0,0 +1,421 @@
+    List<DUResponse.DiskUsage> sortedDuData = new ArrayList<>(duData);
+    // Sort the DU data by size in descending order to compare with the original.
+    sortedDuData.sort(
+        Comparator.comparingLong(DUResponse.DiskUsage::getSize).reversed());
```
Can you pls explain this sorting and verification. As per my understanding, both `duData` and `sortedDuData` will be sorted already, so at this line, are you trying to reverse , then it should be in ascending order and subpaths should not match ? Am I missing something here ?

ArafatKhan2198 replied:
The `verifyOrdering` method is designed to validate that the disk usage data returned from the `nsSummaryEndpoint.getDiskUsage` method is correctly sorted in descending order by size. To check this, it first copies the original data (`duData`) into a new list (sortedDuData), and then sorts this new list in descending order explicitly. By comparing the subpaths of the disk usage objects in both the original and sorted lists, the method confirms that the original data was indeed sorted correctly.

devmadhuu replied:
but original data 'duData' wouldn't be already sorted ?

ArafatKhan2198 replied:
Right if the original is sorted then the ordering will be the same as sortedDuData and hence this test will pass, but if the original duData is not sorted then the ordering with the sortedDuData will not match hence it will result in failure.
That is where the test fails.

devmadhuu replied:
Ok thanks @ArafatKhan2198 for explaining in detail. Its ok to have it this way.

ArafatKhan2198 commented:
> > @devmadhuu @adoroszlai @smitajoshi12
> > Could you please review the latest changes? Here's a quick summary:
> > ```
> > * Switched to Parallel Sorting: To improve performance, we're now using parallel sorting. More details are in the description.
> > 
> > * Added a Toggle for Sorting: There's a new boolean flag to turn sorting on or off.
> > 
> > * Set a Limit of 30 Records: We've added a constant to limit the response to the top 30 records in Disk Usage.
> > ```
> 
> Thanks @ArafatKhan2198 for handling some points. However I am not sure if parallelStreaming always improves performance, in fact rather sometimes, it increases more overhead and may do bad than good. I would like you to have a look [here](https://blogs.oracle.com/javamagazine/post/java-parallel-streams-performance-benchmark).

Thanks a lot, devmadhuu, for the comment and the article! I've read through it carefully and here's my analysis:

**Parallel Streaming concern:**

- Parallel streams introduce overhead for managing multiple threads.
- This overhead can outweigh the benefits of parallel processing for small datasets or simple operations.

After going through the article I can summarise the following ➖

- **Factors affecting performance:**
    - **Data size:** Parallel streams benefit from large datasets where the overhead is justified.
        - This sorting algorithm will be applied to response objects at a single level in the file system hierarchy, which could potentially encompass several hundred thousand items in the worst-case scenario under ideal conditions.
    - **Computation intensity:** Operations involving complex calculations benefit more from parallelization.
        - Sorting is considered a **moderately complex** calculation in the context of parallelization.
    - **Stream source:** Easily splittable sources like arrays perform better in parallel streams,
        - We are using Lists as our source.

adoroszlai commented:
@ArafatKhan2198 @devmadhuu Please omit `@mention` when quoting the [message](https://github.com/apache/ozone/pull/6318#issuecomment-1986286854) that asked for review.  Including it re-subscribes folks mentioned who may have already unsubscribed from the discussion (sorry, I don't have time to review this).

devmadhuu commented:
> > > @devmadhuu @adoroszlai @smitajoshi12
> > > Could you please review the latest changes? Here's a quick summary:
> > > ```
> > > * Switched to Parallel Sorting: To improve performance, we're now using parallel sorting. More details are in the description.
> > > 
> > > * Added a Toggle for Sorting: There's a new boolean flag to turn sorting on or off.
> > > 
> > > * Set a Limit of 30 Records: We've added a constant to limit the response to the top 30 records in Disk Usage.
> > > ```
> > 
> > 
> > Thanks @ArafatKhan2198 for handling some points. However I am not sure if parallelStreaming always improves performance, in fact rather sometimes, it increases more overhead and may do bad than good. I would like you to have a look [here](https://blogs.oracle.com/javamagazine/post/java-parallel-streams-performance-benchmark).
> 
> Thanks a lot, @devmadhuu , for the comment and the article! I've read through it carefully and here's my analysis:
> 
> **Parallel Streaming concern:**
> 
> * Parallel streams introduce overhead for managing multiple threads.
> * This overhead can outweigh the benefits of parallel processing for small datasets or simple operations.
> 
> After going through the article I can summarise the following ➖
> 
> * **Factors affecting performance:**
>   
>   * **Data size:** Parallel streams benefit from large datasets where the overhead is justified.
>     
>     * This sorting algorithm will be applied to response objects at a single level in the file system hierarchy, which could potentially encompass millions of items in the worst-case scenario under ideal conditions.
>   * **Computation intensity:** Operations involving complex calculations benefit more from parallelization.
>     
>     * Sorting is considered a **moderately complex** calculation in the context of parallelization.
>   * **Stream source:** Easily splittable sources like arrays perform better in parallel streams,
>     
>     * We are using Lists as our source.

Do we have any performance measure data  over 1 million records at least with and without parallel streaming. I am emphasizing it because I have experienced , that even with few 10K of records, parallel streaming do bad more than good. So I would suggest to publish some figures of performance with and without parallel streaming at least with 1 million records.

devmadhuu commented:
> @devmadhuu @adoroszlai @smitajoshi12
> 
> Could you please review the latest changes? Here's a quick summary:
> 
> * Switched to Parallel Sorting: To improve performance, we're now using parallel sorting. More details are in the description.
> * Added a Toggle for Sorting: There's a new boolean flag to turn sorting on or off.
> * Set a Limit of 30 Records: We've added a constant to limit the response to the top 30 records in Disk Usage.

Pls check on UI, what is the max limit in dropdown we are setting and using. I think its changed to 10k+. Pls check and confirm.

ArafatKhan2198 commented:
> > > > Could you please review the latest changes? Here's a quick summary:
> > > > ```
> > > > * Switched to Parallel Sorting: To improve performance, we're now using parallel sorting. More details are in the description.
> > > > 
> > > > * Added a Toggle for Sorting: There's a new boolean flag to turn sorting on or off.
> > > > 
> > > > * Set a Limit of 30 Records: We've added a constant to limit the response to the top 30 records in Disk Usage.
> > > > ```
> > > 
> > > 
> > > Thanks @ArafatKhan2198 for handling some points. However I am not sure if parallelStreaming always improves performance, in fact rather sometimes, it increases more overhead and may do bad than good. I would like you to have a look [here](https://blogs.oracle.com/javamagazine/post/java-parallel-streams-performance-benchmark).
> > 
> > 
> > Thanks a lot, @devmadhuu , for the comment and the article! I've read through it carefully and here's my analysis:
> > **Parallel Streaming concern:**
> > 
> > * Parallel streams introduce overhead for managing multiple threads.
> > * This overhead can outweigh the benefits of parallel processing for small datasets or simple operations.
> > 
> > After going through the article I can summarise the following ➖
> > 
> > * **Factors affecting performance:**
> >   
> >   * **Data size:** Parallel streams benefit from large datasets where the overhead is justified.
> >     
> >     * This sorting algorithm will be applied to response objects at a single level in the file system hierarchy, which could potentially encompass millions of items in the worst-case scenario under ideal conditions.
> >   * **Computation intensity:** Operations involving complex calculations benefit more from parallelization.
> >     
> >     * Sorting is considered a **moderately complex** calculation in the context of parallelization.
> >   * **Stream source:** Easily splittable sources like arrays perform better in parallel streams,
> >     
> >     * We are using Lists as our source.
> 
> Do we have any performance measure data over 1 million records at least with and without parallel streaming. I am emphasizing it because I have experienced , that even with few 10K of records, parallel streaming do bad more than good. So I would suggest to publish some figures of performance with and without parallel streaming at least with 1 million records.

Thanks for the comments @devmadhuu tested this out on a cluster with 10 million keys, 
These were the results :- 
```
Sequential sort time: 7657 ms
Parallel sort time: 1279 ms
```

I believe we could got with parallel sort.

devmadhuu commented:
> > > > > Could you please review the latest changes? Here's a quick summary:
> > > > > ```
> > > > > * Switched to Parallel Sorting: To improve performance, we're now using parallel sorting. More details are in the description.
> > > > > 
> > > > > * Added a Toggle for Sorting: There's a new boolean flag to turn sorting on or off.
> > > > > 
> > > > > * Set a Limit of 30 Records: We've added a constant to limit the response to the top 30 records in Disk Usage.
> > > > > ```
> > > > 
> > > > 
> > > > Thanks @ArafatKhan2198 for handling some points. However I am not sure if parallelStreaming always improves performance, in fact rather sometimes, it increases more overhead and may do bad than good. I would like you to have a look [here](https://blogs.oracle.com/javamagazine/post/java-parallel-streams-performance-benchmark).
> > > 
> > > 
> > > Thanks a lot, @devmadhuu , for the comment and the article! I've read through it carefully and here's my analysis:
> > > **Parallel Streaming concern:**
> > > 
> > > * Parallel streams introduce overhead for managing multiple threads.
> > > * This overhead can outweigh the benefits of parallel processing for small datasets or simple operations.
> > > 
> > > After going through the article I can summarise the following ➖
> > > 
> > > * **Factors affecting performance:**
> > >   
> > >   * **Data size:** Parallel streams benefit from large datasets where the overhead is justified.
> > >     
> > >     * This sorting algorithm will be applied to response objects at a single level in the file system hierarchy, which could potentially encompass millions of items in the worst-case scenario under ideal conditions.
> > >   * **Computation intensity:** Operations involving complex calculations benefit more from parallelization.
> > >     
> > >     * Sorting is considered a **moderately complex** calculation in the context of parallelization.
> > >   * **Stream source:** Easily splittable sources like arrays perform better in parallel streams,
> > >     
> > >     * We are using Lists as our source.
> > 
> > 
> > Do we have any performance measure data over 1 million records at least with and without parallel streaming. I am emphasizing it because I have experienced , that even with few 10K of records, parallel streaming do bad more than good. So I would suggest to publish some figures of performance with and without parallel streaming at least with 1 million records.
> 
> Thanks for the comments @devmadhuu tested this out on a cluster with 10 million keys, These were the results :-
> 
> ```
> Sequential sort time: 7657 ms
> Parallel sort time: 1279 ms
> ```
> 
> I believe we could got with parallel sort.

Thanks @ArafatKhan2198 for testing out and publish the figures. This looks promising.

devmadhuu reviewed (commented):
Changes LGTM +1. Pls resolve conflicts.

devmadhuu commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/handlers/BucketEntityHandler.java:
```diff
@@ -142,7 +145,15 @@ public DUResponse getDuResponse(
+    if (sortSubpaths) {
+      // Parallel sort directory/files DU data in descending order of size and returns the top N elements.
+      dirDUData = sortDiskUsageDescendingWithLimit(dirDUData,
+          DISK_USAGE_TOP_RECORDS_LIMIT);
```
Can we make this limit pass as parameter instead of hardcoding here ? Just move one level above.

ArafatKhan2198 replied:
As discussed, we will ignore this comment for this PR and address it in another one.

devmadhuu reviewed (commented):
A minor comment.

devmadhuu reviewed (approved):
Thanks @ArafatKhan2198  for working on this patch. Changes LGTM +1

ArafatKhan2198 commented:
@dombizita Could you please take a final look at it!
I believe we are done and can merge it

dombizita commented on hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/api/TestNSSummaryDiskUsageOrdering.java:
```diff
@@ -0,0 +1,421 @@
+    return reconSCM;
+  }
+
+  private static BucketLayout getBucketLayout() {
+    return BucketLayout.FILE_SYSTEM_OPTIMIZED;
+  }
```
One last quick question: is this only tested with FSO buckets?

ArafatKhan2198 replied:
Thanks for the comment @dombizita 

We've tested this manually for all bucket types. The changes don't really affect the result based on the bucket types, as the changes work directly towards the du response as you can see in the `VolumeEntityHandler` the `sortDiskUsageDescendingWithLimit()` is applied to the response list at the end when the entire response is created. so it does not really matter what the bucket type is.

dombizita replied:
Thanks for explaining it @ArafatKhan2198!

devmadhuu commented:
Thanks @ArafatKhan2198  for working on this patch.
